austrakka/components/sequence/sync/sync_workflow.py

```diff
@@ -0,0 +1,474 @@
+from .constant import DETECTION_DATE_KEY
+from .constant import OBSOLETE_OBJECTS_FILE_KEY
+from .constant import INTERMEDIATE_MANIFEST_FILE_KEY
+from .constant import MANIFEST_KEY
+from .constant import FILE_NAME_ON_DISK_KEY
+from .constant import FILE_NAME_KEY
+from .constant import SEQ_ID_KEY
+from .constant import SEQ_TYPE_KEY
+from .constant import TYPE_KEY
+from .constant import READ_KEY
+from .constant import GROUP_NAME_KEY
+from .constant import BLOB_FILE_PATH_KEY
+from .constant import ORIGINAL_FILE_NAME_KEY
+from .constant import SERVER_SHA_256_KEY
+
+from .constant import MATCH
+from .constant import DOWNLOADED
+from .constant import DRIFTED
+from .constant import FAILED
+from .constant import DONE
+from .constant import MISSING
+from .constant import FASTA_EXT
+from .constant import FASTQ_EXT
+from .constant import GZ_EXT
+from .constant import TRASH_DIR
+
+from ..funcs import _get_seq_data
+
+
+def configure_state_machine() -> StateMachine:
+
+    # Configure the generic state machine with a list of allowed
+    # states and actions (a.k.a transitions). The state machine
+    # uses this information to do a cursory check of your handler
+    # implementations. E.g. if your handler declares that the next
+    # state is 'x' and it is not known, it will raise a runtime
+    # exception and halt rather than potentially corrupting your
+    # data.
+    return StateMachine({
+        SName.PULLING_MANIFEST: State(SName.PULLING_MANIFEST),
+        SName.DONE_PULLING_MANIFEST: State(SName.DONE_PULLING_MANIFEST),
+        SName.ANALYSING: State(SName.ANALYSING),
+        SName.DONE_ANALYSING: State(SName.DONE_ANALYSING),
+        SName.DOWNLOADING: State(SName.DOWNLOADING),
+        SName.DONE_DOWNLOADING: State(SName.DONE_DOWNLOADING),
+        SName.FINALISING: State(SName.FINALISING),
+        SName.DONE_FINALISING: State(SName.DONE_FINALISING),
+        SName.FINALISATION_FAILED: State(SName.FINALISATION_FAILED, is_end_state=True),
+        SName.PURGING: State(SName.PURGING),
+        SName.DONE_PURGING: State(SName.DONE_PURGING),
+        SName.UP_TO_DATE: State(SName.UP_TO_DATE, is_end_state=True),
+    }, {
+        Action.set_state_pulling_manifest: set_state_pulling_manifest,
+        Action.pull_manifest: pull_manifest,
+        Action.set_state_analysing: set_state_analysing,
+        Action.analyse: analyse,
+        Action.set_state_downloading: set_state_downloading,
+        Action.download: download,
+        Action.set_state_finalising: set_state_finalising,
+        Action.finalise: finalise,
+        Action.set_state_purging: set_state_purging,
+        Action.purge: purge,
+        Action.set_state_up_to_date: set_state_up_to_date
+    })
+
+
+def set_state_pulling_manifest(sync_state: dict):
+    sync_state[CURRENT_STATE_KEY] = SName.PULLING_MANIFEST
+    sync_state[CURRENT_ACTION_KEY] = Action.pull_manifest
+
+
+def pull_manifest(sync_state: dict):
+    logger.info(f'Started: {Action.pull_manifest}')
```

Comment:
Could these start/finish log statements be moved into the state machine itself?

Reply:
I think it's closer to the source of truth if it's done inside the action, which already logs about other events.